### PR TITLE
Display 3d graphics in section drawing views

### PR DESCRIPTION
**Pull request by pmconne**
If the `displaySpatialView` JSON property of the associated SectionDrawing is `true`, we display the spatial view in the context of the drawing view.
I didn't want to add yet more code to the already-bloated ViewState.ts so split classes out into DrawingViewState.ts and SpatialViewState.ts.
Info about the section drawing is included in ViewStateProps if relevant.
Moved guts of Viewport.changeViewedModel2d() to ViewState2d and cleaned it up.
Added ViewState.toProps() as the inverse of createFromProps().

TODO in separate PR:
I noticed that the Viewports I use to select tiles for the spatial view attachment do not get disposed of - they remain registered with TileAdmin. The same is true of the Viewports used for sheet view attachments; that is more of an issue because those viewports may allocate (and never free) WebGL resources if they are displaying a perspective view.
We need to make sure they get disposed of. Probably the view should be informed when it becomes attached to or detached from a Viewport. That would also help with making sure that changes made to the view, display style, etc are propagated to the Viewport so the caller doesn't need to explicitly invalidate the scene etc or to remember to call the versions of those APIs on Viewport that invalidate things for him.

### CODE REVIEW

**Comment by pmconne** [comment]
I'm adding a few more tests.

File: core/common/src/ElementProps.ts [hunk 1]
```diff
@@ -12,7 +12,16 @@ import {
...
+import {
+  AngleProps,
+  ClipVectorProps,
+  LowAndHighXY,
+  LowAndHighXYZ,
+  TransformProps,
+  XYProps,
+  XYZProps,
+  YawPitchRollProps,
+} from "@bentley/geometry-core";
```
Comment (scsewall) [inline core/common/src/ElementProps.ts:24]: Fix import formatting?

**Review by scsewall — APPROVED** [review]

File: core/common/src/ElementProps.ts (hunk 1 above)
Comment (pmconne) [reply core/common/src/ElementProps.ts:17]: We need to start using prettier instead of wasting time on silly stuff.

**Comment by pmconne** [comment]
@kabentley please take a look at the ViewState and Viewport changes. Also, any objections to the proposed TODO in PR description?

**Comment by pmconne** [comment]
@mgooding since most of the display team is out for the holidays would you mind reviewing?

**Review by mgooding — APPROVED** [review]

File: core/frontend/src/ViewState.ts [hunk 2]
```diff
@@ -256,6 +257,15 @@ export abstract class ViewState extends ElementState {
    */
   public static createFromProps(_props: ViewStateProps, _iModel: IModelConnection): ViewState | undefined { return undefined; }
 
+  /** Serialize this ViewState as a set of properties that can be used to recreate it via [[ViewState.createFromProps]]. */
+  public toProps(): ViewStateProps {
```
Comment (kabentley) [inline core/frontend/src/ViewState.ts:261]: Don't we usually call this toJSON?

Comment (pmconne) [reply core/frontend/src/ViewState.ts:261]: `toProps` is the inverse of `createFromProps` as `toJSON` is the inverse of `fromJSON`.

File: core/frontend/src/ViewState.ts [hunk 3]
```diff
@@ -364,15 +374,15 @@ export abstract class ViewState extends ElementState {
   public abstract get details(): ViewDetails;
 
   /** Returns true if this ViewState is-a [[ViewState3d]] */
-  public is3d(): this is ViewState3d { return this instanceof ViewState3d; }
```
Comment (kabentley) [inline core/frontend/src/ViewState.ts:367]: should these be abstract?

File: core/frontend/src/DrawingViewState.ts [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    super(props, iModel, categories, displayStyle);
+    if (categories instanceof DrawingViewState) {
+      this._viewedExtents = categories._viewedExtents.clone();
+      this._modelLimits = { ...categories._modelLimits };
+      this._attachment = categories._attachment?.clone();
+      this._sectionDrawingProps = categories._sectionDrawingProps;
+    } else {
+      this._viewedExtents = extents;
+      this._modelLimits = { min: Constant.oneMillimeter, max: 10 * extents.maxLength() };
+      this._sectionDrawingProps = sectionDrawing ?? defaultSectionDrawingProps;
```
Comment (kabentley) [inline core/frontend/src/DrawingViewState.ts:264]: what's the difference between undefined and defaultSectionDrawingProps?

**Comment by kabentley** [comment]
Do you propose to put methods on ViewState "onAttached/onDetached" or something? Is it possible for a ViewState to be attached to more than one Viewport?

File: core/frontend/src/ViewState.ts (hunk 3 above)
Comment (kabentley) [reply core/frontend/src/ViewState.ts:367]: or:
```ts
public abstract is3d(): boolean;
public is2d(): boolean { return !this.is3d(); }
```

**Comment by pmconne** [comment]
> Do you propose to put methods on ViewState "onAttached/onDetached" or something? Is it possible for a ViewState to be attached to more than one Viewport?

ViewState will have methods invoked by Viewport when changeView() is used to attach or detach the view.
It has always been technically possible for the same ViewState to be associated with multiple Viewports, or the same DisplayStyle with multiplie ViewStates. It has never been a good idea, because the Viewport modifies the ViewState's state.

File: core/frontend/src/ViewState.ts (hunk 3 above)
Comment (pmconne) [reply core/frontend/src/ViewState.ts:367]: Probably.

File: core/frontend/src/DrawingViewState.ts (hunk 4 above)
Comment (pmconne) [reply core/frontend/src/DrawingViewState.ts:264]: `undefined` means we don't know what the section drawing props are. This occurs as a result of `changeViewedModel`.

**Comment by kabentley** [comment]
> It has never been a good idea, because the Viewport modifies the ViewState's state.

OK. Maybe in your changes you can throw if a ViewState is attached to multiple Viewports, with a "clone your ViewState" message.

File: core/frontend/src/ViewState.ts (hunk 3 above)
Comment (pmconne) [reply core/frontend/src/ViewState.ts:367]: Pushed that change.

Comment (pmconne) [reply core/frontend/src/ViewState.ts:367]: Please resolve comments if changes/responses satisfactory.

**Pull request merged**